{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Abstracts Using a Large Language Model (LLM)\n",
    "\n",
    "## Introduction\n",
    "\n",
    "In this Jupyter Notebook, we will perform an in-depth analysis of abstracts extracted from a CSV file using a Large Language Model (LLM). The goal of this analysis is to leverage the capabilities of LLMs to extract meaningful insights, identify key themes, and perform various natural language processing (NLP) tasks on the abstracts.\n",
    "\n",
    "### Objectives\n",
    "\n",
    "- **Data Loading**: Import and preprocess abstracts from a CSV file.\n",
    "- **Text Analysis**: Utilize LLMs to analyze the content of the abstracts.\n",
    "\n",
    "### Tools and Libraries\n",
    "\n",
    "- **LangChain**: To interface with the LLM.\n",
    "\n",
    "### Workflow\n",
    "\n",
    "1. **Data Import**: Load the CSV file containing the abstracts.\n",
    "3. **LLM Integration**: Use the LLM to perform various NLP tasks.\n",
    "\n",
    "By the end of this notebook, you will have a comprehensive understanding of how to use LLMs for analyzing textual data and extracting valuable insights from scientific abstracts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. **Data Import**: Load the CSV file containing the abstracts.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from langchain_community.llms import Ollama\n",
    "from langchain_core.output_parsers import JsonOutputParser\n",
    "from langchain_core.prompts import PromptTemplate\n",
    "from downloader import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'10.1038/s41586-020-2746-2': {'abstract': 'The genetic circuits that allow cancer cells to evade destruction by the host immune system remain poorly understood13. Here, to identify a phenotypically robust core set of genes and pathways that enable cancer cells to evade killing mediated by cytotoxic Tlymphocytes (CTLs), we performed genome-wide CRISPR screens across a panel of genetically diverse mouse cancer cell lines that were cultured in the presence of CTLs. We identify a core set of 182genes across these mouse cancer models, the individual perturbation of which increases either the sensitivity or the resistance of cancer cells to CTL-mediated toxicity. Systematic exploration of our dataset using genetic co-similarity reveals the hierarchical and coordinated manner in which genes and pathways act in cancer cells to orchestrate their evasion of CTLs, and shows that discrete functional modules that control the interferon response and tumour necrosis factor (TNF)-induced cytotoxicity are dominant sub-phenotypes. Our data establish a central role for genes that were previously identified as negative regulators of the type-II interferon response (for example, Ptpn2, Socs1 and Adar1) in mediating CTL evasion, and show that the lipid-droplet-related gene Fitm2 is required for maintaining cell fitness after exposure to interferon-? (IFN?). In addition, we identify the autophagy pathway as a conserved mediator of the evasion of CTLs by cancer cells, and show that this pathway is required to resist cytotoxicity induced by the cytokines IFN? andTNF. Through the mapping of cytokine- and CTL-based genetic interactions, together with in vivo CRISPR screens, we show how the pleiotropic effects of autophagy control cancer-cell-intrinsic evasion of killing by CTLs and we highlight the importance of these effects within the tumour microenvironment. Collectively, thesedata expand our knowledge of the genetic circuits that are involved in the evasion of the immune system by cancer cells, and highlight genetic interactions that contribute to phenotypes associated with escape from killing by CTLs.'}, '10.1186/s12885-016-2539-z': {'abstract': 'Background: Patients with clear cell renal cell carcinoma (ccRCC) have few therapeutic options, as ccRCC is unresponsive to chemotherapy and is highly resistant to radiation. Recently targeted therapies have extended progression-free survival, but responses are variable and no significant overall survival benefit has been achieved. Commercial ccRCC cell lines are often used as model systems to develop novel therapeutic approaches, but these do not accurately recapitulate primary ccRCC tumors at the genomic and transcriptional levels. Furthermore, ccRCC exhibits significant intertumor genetic heterogeneity, and the limited cell lines available fail to represent this aspect of ccRCC. Our objective was to generate accurate preclinical in vitro models of ccRCC using tumor tissues from ccRCC patients. Methods: ccRCC primary single cell suspensions were cultured in fetal bovine serum (FBS)-containing media or defined serum-free media. Established cultures were characterized by genomic verification of mutations present in the primary tumors, expression of renal epithelial markers, and transcriptional profiling. Results: The apparent efficiency of primary cell culture establishment was high in both culture conditions, but genotyping revealed that the majority of cultures contained normal, not cancer cells. ccRCC characteristically shows biallelic loss of the von Hippel Lindau (VHL) gene, leading to accumulation of hypoxia-inducible factor (HIF) and expression of HIF target genes. Purification of cells based on expression of carbonic anhydrase IX (CA9), a cell surface HIF target, followed by culture in FBS enabled establishment of ccRCC cell cultures with an efficiency of >80 %. Culture in serum-free conditions selected for growth of normal renal proximal tubule epithelial cells. Transcriptional profiling of ccRCC and matched normal cell cultures identified up- and down-regulated networks in ccRCC and comparison to The Cancer Genome Atlas confirmed the clinical validity of our cell cultures. Conclusions: The ability to establish primary cultures of ccRCC cells and matched normal kidney epithelial cells from almost every patient provides a resource for future development of novel therapies and personalized medicine for ccRCC patients.'}, '10.1038/nmeth.2895': {'abstract': 'Cell signaling, one of key processes in both normal cellular function and disease, is coordinated by numerous interactions between membrane proteins that change in response to stimuli. We present a split ubiquitin-based method for detection of integral membrane protein-protein interactions (PPIs) in human cells, termed mammalian-membrane two-hybrid assay (MaMTH). We show that this technology detects stimulus (hormone or agonist)-dependent and phosphorylation-dependent PPIs. MaMTH can detect changes in PPIs conferred by mutations such as those in oncogenic ErbB receptor variants or by treatment with drugs such as the tyrosine kinase inhibitor erlotinib. Using MaMTH as a screening assay, we identified CRKII as an interactor of oncogenic EGFR(L858R) and showed that CRKII promotes persistent activation of aberrant signaling in non-small cell lung cancer cells. MaMTH is a powerful tool for investigating the dynamic interactomes of human integral membrane proteins.  2014 Nature America, Inc.'}, '10.1016/j.celrep.2012.09.016': {'abstract': 'The pentaspan membrane glycoprotein CD133 marks lineage-specific cancer progenitor cells and is associated with poor prognosis in a number of tumor types. Despite its utility as a cancer progenitor cell marker, CD133 protein regulation and molecular function remain poorly understood. We find that the deacetylase HDAC6 physically associates with CD133 to negatively regulate CD133 trafficking down the endosomal-lysosomal pathway for degradation. We further demonstrate that CD133, HDAC6, and the central molecule of the canonical Wnt signaling pathway, ?-catenin, can physically associate as a ternary complex. This association stabilizes ?-catenin via HDAC6 deacetylase activity, which leads to activation of ?-catenin signaling targets. Downregulation of either CD133 or HDAC6 results in increased ?-catenin acetylation and degradation, which correlates with decreased proliferation in vitro and tumor xenograft growth in vivo. Given that CD133 marks progenitor cells in a wide range of cancers, targeting CD133 may be a means to treat multiple cancer types'}, '10.1158/1535-7163.MCT-09-0495': {'abstract': 'Evasion of death receptor ligand-induced apoptosis represents an important contributor to cancer development and progression. Therefore, molecules that restore sensitivity to death receptor stimuli would be important tools to better understand this biological pathway and potential leads for therapeutic adjuncts. Previously, the small-molecule 4-(4-chloro-2-methylphenoxy)-N- hydroxybutanamide (that we propose be named droxinostat) was identified as a chemical sensitizer to death receptor stimuli, decreasing the expression of the caspase-8 inhibitor FLIP. However, the direct targets of droxinostat were unknown. To better understand the mechanism of action of droxinostat and highlight new strategies to restore sensitivity to death receptor ligands, we analyzed changes in gene expression using the Connectivity Map after treating cells with droxinostat. Changes in gene expression after droxinostat treatment resembled changes observed after treatment with histone deacetylase (HDAC) inhibitors. Therefore, we examined the effects of droxinostat on HDAC activity and showed that it selectively inhibited HDAC3, HDAC6, and HDAC8 and that inhibition of these HDACs was functionally important for its ability to sensitize cells to death ligands. Thus, we have identified a selective HDAC inhibitor and showed that selective HDAC inhibition sensitizes cells to death ligands, thereby highlighting a new mechanism to overcome resistance to death receptor ligands. 2010 AACR.'}, '10.1182/blood-2009-07-231191': {'abstract': 'The proteasomal pathway of protein degradation involves 2 discrete steps: ubiquitination and degradation. Here, we evaluated the effects of inhibiting the ubiquitination pathway at the level of the ubiquitin-activating enzyme UBA1 (E1). By immunoblotting, leukemia cell lines and primary patient samples had increased protein ubiquitination. Therefore, we examined the effects of genetic and chemical inhibition of the E1 enzyme. Knockdown of E1 decreased the abundance of ubiquitinated proteins in leukemia and myeloma cells and induced cell death. To further investigate effects of E1 inhibition in malignancy, we discovered a novel small molecule inhibitor, 3,5-dioxopyrazolidine compound, 1-(3-chloro-4-fluorophenyl)-4-[(5-nitro-2-furyl)methylene]-3, 5-pyrazolidinedione (PYZD-4409). PYZD-4409 induced cell death in malignant cells and preferentially inhibited the clonogenic growth of primary acute myeloid leukemia cells compared with normal hematopoietic cells. Mechanistically, genetic or chemical inhibition of E1 increased expression of E1 stress markers. Moreover, BI-1 overexpression blocked cell death after E1 inhibition, suggesting ER stress is functionally important for cell death after E1 inhibition. Finally, in a mouse model of leukemia, intraperitoneal administration of PYZD-4409 decreased tumor weight and volume compared with control without untoward toxicity. Thus, our work highlights the E1 enzyme as a novel target for the treatment of hematologic malignancies.  2010 by The American Society of Hematology.'}, '10.1038/nmeth.3472': {'abstract': \"Antibodies are used in multiple cell biology applications, but there are no standardized methods to assess antibody quality - an absence that risks data integrity and reproducibility. We describe a mass spectrometry-based standard operating procedure for scoring immunoprecipitation antibody quality. We quantified the abundance of all the proteins in immunoprecipitates of 1,124 new recombinant antibodies for 152 chromatin-related human proteins by comparing normalized spectral abundance factors from the target antigen with those of all other proteins. We validated the performance of the standard operating procedure in blinded studies in five independent laboratories. Antibodies for which the target antigen or a member of its known protein complex was the most abundant protein were classified as 'IP gold standard'. This method generates quantitative outputs that can be stored and archived in public databases, and it represents a step toward a platform for community benchmarking of antibody quality.\"}, '10.1074/jbc.M113.469262': {'abstract': 'Ubiquitin-specific protease 7 (USP7) is a deubiquitinating enzyme found in all eukaryotes that catalyzes the removal of ubiquitin from specific target proteins. Here, we report that UbE2E1, an E2 ubiquitin conjugation enzyme with a unique N-terminal extension, is a novel USP7-interacting protein. USP7 forms a complex with UbE2E1 in vitro and in vivo through theASTSUSP7 binding motif within its N-terminal extension in an identical manner with other known USP7 binding proteins. We show that USP7 attenuates UbE2E1-mediated ubiquitination, an effect that requires the N-terminal ASTS sequence of UbE2E1 as well as the catalytic activity of USP7. Additionally, USP7 is critical in maintaining the steady state levels of UbE2E1 in cells. This study reveals a new cellular mechanism that couples the opposing activities of the ubiquitination machinery and a deubiquitinating enzyme to maintain and modulate the dynamic balance of the ubiquitin-proteasome system.  2013 by The American Society for Biochemistry and Molecular Biology, Inc.'}, '10.1016/j.molcel.2005.02.029': {'abstract': 'USP7/HAUSP is a key regulator of p53 and Mdm2 and is targeted by the Epstein-Barr nuclear antigen 1 (EBNA1) protein of Epstein-Barr virus (EBV). We have determined the crystal structure of the p53 binding domain of USP7 alone and bound to an EBNA1 peptide. This domain is an eight-stranded ? sandwich similar to the TRAF-C domains of TNF-receptor associated factors, although the mode of peptide binding differs significantly from previously observed TRAF-peptide interactions in the sequence (DPGEGPS) and the conformation of the bound peptide. NMR chemical shift analyses of USP7 bound by EBNA1 and p53 indicated that p53 binds the same pocket as EBNA1 but makes less extensive contacts with USP7. Functional studies indicated that EBNA1 binding to USP7 can protect cells from apoptotic challenge by lowering p53 levels. The data provide a structural and conceptual framework for understanding how EBNA1 might contribute to the survival of Epstein-Barr virus-infected cells. Copyright 2005 by Elsevier Inc.'}, '10.1074/jbc.M307200200': {'abstract': 'USP7 or HAUSP is a ubiquitin-specific protease in human cells that regulates the turnover of p53 and is bound by at least two viral proteins, the ICP0 protein of herpes simplex type 1 and the EBNA1 protein of Epstein-Barr virus. We have overexpressed and purified USP7 and shown that the purified protein is monomeric and is active for cleaving both a linear ubiquitin substrate and conjugated ubiquitin on EBNA1. Using partial proteolysis of USP7 coupled with matrix-assisted laser desorption ionization time-of-flight mass spectrometry, we showed that USP7 comprises four structural domains; an N-terminal domain known to bind p53, a catalytic domain, and two C-terminal domains. By passing a mixture of USP7 domains over EBNA1 and ICP0 affinity columns, we showed that the N-terminal p53 binding domain was also responsible for the EBNA1 interaction, while the ICP0 binding domain mapped to a C-terminal domain between amino acids 599-801. Tryptophan fluorescence assays showed that an EBNA1 peptide mapping to residues 395-450 was sufficient to bind the USP7 N-terminal domain and did so with a dissociation constant of 0.9-2 ?M, whereas p53 peptides spanning the USP7-binding region gave dissociation constants of 9-17 ?M in the same assay. In keeping with these relative affinities, gel filtration analyses of the complexes showed that the EBNA1 peptide efficiently competed with the p53 peptide for USP7 binding, suggesting that EBNA1 could affect p53 function in vivo by competing for USP7.'}, '10.1038/nsmb1067': {'abstract': 'The ubiquitin-specific protease, USP7, has key roles in the p53 pathway whereby it stabilizes both p53 and MDM2. We show that the N-terminal domain of USP7 binds two closely spaced 4-residue sites in both p53 and MDM2, falling between p53 residues 359-367 and MDM2 residues 147-159. Cocrystal structures with USP7 were determined for both p53 peptides and for one MDM2 peptide. These peptides bind the same surface of USP7 as Epstein-Barr nuclear antigen-1, explaining the competitive nature of the interactions. The structures and mutagenesis data indicate a preference for a P/AXXS motif in peptides that bind USP7. Contacts made by serine are identical and crucial for all peptides, and Trp165 in the peptide-binding pocket of USP7 is also crucial. These results help to elucidate the mechanism of substrate recognition by USP7 and the regulation of the p53 pathway.  2006 Nature Publishing Group.'}, '10.1016/j.jmb.2010.08.017': {'abstract': 'Ubiquitin-specific protease 7 (USP7) catalyzes the deubiquitination of several substrate proteins including p53 and Hdm2. We have previously shown that USP7, and more specifically its amino-terminal domain (USP7-NTD), interacts with distinct regions on p53 and Hdm2 containing P/AxxS motifs. The ability of USP7 to also deubiquitinate and control the turnover of HdmX was recently demonstrated. We utilized a combination of biochemistry and structural biology to identify which domain of USP7 interacts with HdmX as well as to identify regions of HdmX that interact with USP7. We showed that USP7-NTD recognized two of six P/AxxS motifs of HdmX (8AQCS11 and 398AHSS401). The crystal structure of the USP7-NTD:HdmXAHSS complex was determined providing the molecular basis of complex formation between USP7-NTD and the HdmXAHSS peptide. The HdmX peptide interacted within the same residues of USP7-NTD as previously demonstrated with p53, Hdm2, and EBNA1 peptides. We also identified an additional site on Hdm2 (397PSTS400) that interacts with USP7-NTD and determined the crystal structure of this complex. Finally, analysis of USP7-interacting peptides on filter arrays confirmed the importance of the serine residue at the fourth position for the USP7-NTD interaction and showed that phosphorylation of serines within the binding sequence prevents this interaction. These results lead to a better understanding of the mechanism of substrate recognition by USP7-NTD.  2010 Elsevier Ltd.'}, '10.1074/jbc.M414421200': {'abstract': 'A combination of structural, biochemical, and genetic studies in model organisms was used to infer a cellular role for the human protein (SBDS) responsible for Sliwachman-Bodian-Diamond syndrome. The crystal structure of the SBDS homologue in Archaeoglobus fulgidus, AF0491, revealed a three domain protein. The N-terminal domain, which harbors the majority of disease-linked mutations, has a novel three-dimensional fold. The central domain has the common winged helix-turn-helix motif, and the C-terminal domain shares structural homology with known KNA-binding domains. Proteomic analysis of the SBDS sequence homologue in Saccharomyces cerevisiae, YLR022C, revealed an association with over 20 proteins involved in ribosome biosynthesis, NMR structural genomics revealed another yeast protein, YHR087W, to be a structural homologue of the AF0491 N-terminal domain. Sequence analysis confirmed them as distant sequence homologues, therefore related by divergent evolution. Synthetic genetic array analysis of YHR087W revealed genetic interactions with proteins involved in RNA and rRNA processing including Mdm20/Nat3, Nsr1, and Npl3. Our observations, taken together with previous reports, support the conclusion that SBDS and its homologues play a role in RNA metabolism.  2005 by The American Society for Biochemistry and Molecular Biology, Inc.'}, '10.1016/j.surfcoat.2020.126648': {'abstract': 'The formation of mud-cracking networks in coatings and thin films is a widely observed phenomenon, particularly in Cr-containing electrodeposited materials. Here we have documented and described aspects of mud-cracking specific to Cr-containing electrodeposited alloys made from trivalent Cr electrolytes. In addition to describing the different crack patterns observed in these materials and their relationship to the electroplating parameters, thickness, and composition of the coatings, we also investigated a never-before-reported phenomenon of post-deposition time-dependent cracking. Our analysis shows that cracks continue to form after the electroplating procedure has been completed, generally saturating 2 h after synthesis, but can continue for days afterwards. We theorize that the behaviour is related to the state of oxygen in these samples, which can sometimes reach beyond 30 at.%. It is believed that the breakdown of Cr-hydroxides and Cr-hydrides drives the increase in internal stresses that result in continued cracking. As a result, this time-dependent cracking phenomenon may impact the suitability of certain Cr-alloy electrodeposits in some applications.'}, '10.1002/adfm.200701441': {'abstract': 'Optically active nanostructures comprised of polymer multilayer microspheres surrounding semiconductor quantum dots are studied. Comparative analysis of the optical properties of the microspheres with four alternating layers with different refractive indices is performed. Photoluminescence and extinction characteristics of the microspheres are determined by the dimensions, internal structure, and the average refractive index of the particles.  2008 WILEY-VCH Verlag GmbH & Co. KGaA.'}, '10.1002/1616-3028(20020101)12:1<71::AID-ADFM71>3.0.CO;2-I': {'abstract': 'Barium titanate inverted opals with powder and film morphologies were synthesized from barium ethoxide and titanium isopropoxide in the interstitial spaces of a polystyrene opal. This procedure involves infiltration of precursors into the interstices of the polystyrene opal template followed by hydrolytic polycondensation of the precursors to amorphous barium titanate and removal of the polystyrene opal by solvent extraction or calcination. In-situ variable temperature powder X-ray diffraction and micro-Raman spectroscopy allow one to observe the thermally induced transformation of the as-synthesized amorphous barium titanate inverted opal to the nanocrystalline form. In this way, a nanocrystalline barium titanate inverted opal can be engineered as either the cubic or tetragonal polymorph. Control of this process is key to the practical realization of a room-temperature stable ferroelectric barium titanate inverted opal that can be thermally tuned through the ferroelectric-paraelectric transition around the Curie temperature. Optical characterization demonstrated photonic crystal behavior of the inverted barium titanate opals and results were in good agreement with photonic band structure calculations. The synthesis of optical quality ferroelectric barium titanate inverted opals provides an opportunity to electrically and optically engineer the photonic band structure and the possibility of developing tunable three-dimensional photonic crystal devices.'}, '10.1139/V09-172': {'abstract': 'We studied the optical properties of hybrid multilayer microspheres (HMMs). We prepared multilayer particles with alternating radial layers of gold nanoparticles (NPs) and poly(methyl methacrylate) and achieved control over particle size, the thickness of particle layers, and the surface coverage of gold NPs. We showed the ability to tune the spectral characteristics of the HMMs, which was based on frequency dispersion of the dielectric constant of polymer-metal NP structures. Good agreement between experimental and theoretical extinction properties of the HMMs was obtained. In comparison with multilayer structures synthesized solely from polymers, hybrid multilayer microspheres have a larger refractive layer contrast between adjacent layers, which is important for practical applications in optoelectronic devices.'}, '10.1088/0957-4484/20/29/295203': {'abstract': 'For various types of ensembles of metal nanorods, the frequencies of longitudinal and transverse plasmons were calculated and correlations between the plasmon frequency shifts and the topology of nanorod arrays were found. The theoretical predictions were compared with the experimentally determined optical absorption in arrays of polymer-terminated Au nanorods obtained by self-assembly in selective solvents.  2009 IOP Publishing Ltd.'}, '10.7554/eLife.06011': {'abstract': 'Adipose tissue is crucial for the maintenance of energy and metabolic homeostasis and its deregulation can lead to obesity and type II diabetes (T2D). Using gene disruption in the mouse, we discovered a function for a RhoA-specific guanine nucleotide exchange factor PDZ-RhoGEF (Arhgef11) in white adipose tissue biology. While PDZ-RhoGEF was dispensable for a number of RhoA signaling-mediated processes in mouse embryonic fibroblasts, including stress fiber formation and cell migration, its deletion led to a reduction in their proliferative potential. On a whole organism level, PDZ-RhoGEF deletion resulted in an acute increase in energy expenditure, selectively impaired early adipose tissue development and decreased adiposity in adults. PDZ- RhoGEF-deficient mice were protected from diet-induced obesity and T2D. Mechanistically, PDZ- RhoGEF enhanced insulin/IGF-1 signaling in adipose tissue by controlling ROCK-dependent phosphorylation of the insulin receptor substrate-1 (IRS-1). Our results demonstrate that PDZ- RhoGEF acts as a key determinant of mammalian metabolism and obesity-associated pathologies.'}}\n"
     ]
    }
   ],
   "source": [
    "def read_csv(file_path, num_lines):\n",
    "    with open(file_path, 'r', encoding=\"utf8\", errors='ignore') as file:\n",
    "        reader = csv.reader(file)\n",
    "        dico = {}\n",
    "        for i, row in enumerate(reader):\n",
    "            if i >= num_lines:\n",
    "                break\n",
    "            if i == 0:\n",
    "                continue\n",
    "            dico[row[17]] = {\"abstract\": row[4]} \n",
    "            \"\"\"\n",
    "            try:\n",
    "               \n",
    "                downloader = Downloader(row[17], 'doi', f\"pdfs/{row[17]}.pdf\")\n",
    "                downloader.download()\n",
    "                dico[row[17]].update({\"pdf\": True})\n",
    "            except Exception as e:\n",
    "                dico[row[17]].update({\"pdf\": False})\n",
    "                print(e)\n",
    "          \"\"\"      \n",
    "        print(dico)\n",
    "        return dico\n",
    "\n",
    "file_path = \"all(Scholarly & creative works_Obje).csv\"\n",
    "num_lines = 20\n",
    "dico = read_csv(file_path, num_lines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download incomplete.    doi = 10.1038/s41586-020-2746-2\n",
      "access failed.    doi = 10.1038/nmeth.2895\n"
     ]
    }
   ],
   "source": [
    "# Import necessary libraries\n",
    "from selenium import webdriver  # For automating web browser interaction\n",
    "import time   # For adding delays in the code execution\n",
    "import os     # For file and directory manipulation\n",
    "import random # For generating random numbers\n",
    "\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.service import Service  # Import Service\n",
    "import time\n",
    "import os\n",
    "import random\n",
    "\n",
    "# --------------------- Section 1: Sci-Hub article downloader ---------------------\n",
    "# Function to automate downloading of articles from Sci-Hub using DOIs\n",
    "def scihub_get(doi):\n",
    "    chromeOptions = webdriver.ChromeOptions()  # Set Chrome options for Selenium\n",
    "    prefs = {\"download.default_directory\" : \"/home/youssef/clone/spock/scholarly_creative_work/pdfs\"}  # Your preferred download directory\n",
    "    chromeOptions.add_experimental_option(\"prefs\", prefs)\n",
    "    \n",
    "    # Specify the path to the Chrome driver\n",
    "    chrome_driver_path = \"/usr/bin/chromedriver\"  # Update this with your actual path\n",
    "    service = Service(chrome_driver_path)  # Create a Service object for ChromeDriver\n",
    "    \n",
    "    # Initialize Chrome WebDriver using Service\n",
    "    wd = webdriver.Chrome(service=service, options=chromeOptions)\n",
    "    \n",
    "    # List of Sci-Hub domains to randomly select from\n",
    "    scihub = ['https://sci-hub.ru/', 'https://sci-hub.st/', 'https://sci-hub.se/']\n",
    "    root = scihub[random.randint(0, 2)]  # Select a random Sci-Hub domain\n",
    "    \n",
    "    # Search for the article using its DOI\n",
    "    wd.get(root + doi)  # Open the Sci-Hub page for the DOI\n",
    "    time.sleep(1)  # Pause to let the page load\n",
    "    \n",
    "    try:\n",
    "        # Attempt to click the download button\n",
    "        b = wd.find_element('xpath','//*[@id=\"buttons\"]/button')\n",
    "        b.click()\n",
    "        flag = True\n",
    "        time.sleep(20)  # Wait for the file to download\n",
    "    except:\n",
    "        # Print error message if the article access fails\n",
    "        print('access failed.    doi = ' + doi)\n",
    "        flag = False\n",
    "        time.sleep(5)  # Shorter wait before next attempt\n",
    "    \n",
    "    wd.quit()  # Close the WebDriver\n",
    "    return flag  # Return whether the download was successful\n",
    "\n",
    "# --------------------- Section 2: Rename downloaded files (using DOI only) ---------------------\n",
    "# Function to rename downloaded files based only on DOI\n",
    "def rename_file(doi):\n",
    "    time.sleep(1)  # Wait to ensure the file is downloaded\n",
    "    path = '/home/youssef/clone/spock/scholarly_creative_work/pdfs/'  # Your preferred download directory\n",
    "    \n",
    "    # List all files in the download directory\n",
    "    dir_list = os.listdir(path)\n",
    "    \n",
    "    # Check if any files were downloaded\n",
    "    if len(dir_list) > 0:\n",
    "        found = 0\n",
    "        for file in dir_list:\n",
    "            if file[0:3] != 'No_':  # Skip already renamed files\n",
    "                found = 1  # File found\n",
    "                break\n",
    "\n",
    "        if found == 0:  # If no new file is found\n",
    "            print('download failed.    doi = ' + doi)\n",
    "        else:  # If a new file is found\n",
    "            l = file.split('.')\n",
    "            if l[len(l) - 1] != \"pdf\":  # Check if the file was fully downloaded\n",
    "                print('download incomplete.    doi = ' + doi)\n",
    "            else:\n",
    "                # Rename the file based only on DOI\n",
    "                old = path + '/' + file\n",
    "                doi_cleaned = doi.replace(\"/\", \"_\")  # Replace \"/\" in DOI with \"_\" to avoid issues in file name\n",
    "                new = path + '/No_' + doi_cleaned + '.pdf'\n",
    "                os.rename(old, new)  # Rename the file\n",
    "\n",
    "# --------------------- Section 3: Article download manager ---------------------\n",
    "# Function to handle the entire process for downloading and renaming articles\n",
    "def article_get(doi):\n",
    "    # If the article is successfully downloaded, rename the file\n",
    "    if scihub_get(doi):\n",
    "        rename_file(doi)\n",
    "\n",
    "# --------------------- Section 4: Download articles based on dictionary ---------------------\n",
    "# Assuming 'dico' is your dictionary where DOIs are the keys\n",
    "for doi in dico.keys():\n",
    "    article_get(doi)\n",
    "\n",
    "# --------------------- Section 5: Retry for missing articles ---------------------\n",
    "# Check for articles that were not downloaded and try again\n",
    "path = '/home/youssef/clone/spock/scholarly_creative_work/pdfs/'  # Your preferred download directory\n",
    "dir_list = os.listdir(path)\n",
    "\n",
    "for doi in dico.keys():\n",
    "    doi_cleaned = doi.replace(\"/\", \"_\")  # Replace \"/\" in DOI for file name\n",
    "    filename = 'No_' + doi_cleaned + '.pdf'\n",
    "    \n",
    "    if filename in dir_list:\n",
    "        continue  # Skip if the file already exists\n",
    "    else:\n",
    "        article_get(doi)\n",
    "\n",
    "# --------------------- Section 6: Output missing articles ---------------------\n",
    "# Output information about articles that are still missing after retry\n",
    "count = 0\n",
    "\n",
    "for doi in dico.keys():\n",
    "    doi_cleaned = doi.replace(\"/\", \"_\")  # Replace \"/\" in DOI for file name\n",
    "    filename = 'No_' + doi_cleaned + '.pdf'\n",
    "    \n",
    "    if filename in dir_list:\n",
    "        continue  # Skip if the file already exists\n",
    "    else:\n",
    "        # Print details of missing articles\n",
    "        print(filename + \"    doi: \" + doi)\n",
    "        count += 1\n",
    "\n",
    "# Print the total number of missing articles\n",
    "print(str(count) + \" articles missing in total.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LLM Analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_topic(abstract:str):\n",
    "    \n",
    "    Llm = Ollama(model='llama3', temperature=0.2)\n",
    "    \n",
    "    \n",
    "    if abstract is None:\n",
    "        raise ValueError(\"Abstract is required\")\n",
    "    \n",
    "    parser = JsonOutputParser()\n",
    "    \n",
    "    #AI / accelerated materials discovery / SDLs / autonomous labs / high-throughput experimentation / high-throughput DFT\n",
    "    \n",
    "    topics = [\"Machine Learning\", \"Batteries\", \"AI\", \"accelerated materials discovery\", \"Self Driving Labs\", \"autonomous labs\", \"high-throughput experimentation\", \"high-throughput DFT\"]\n",
    "    \n",
    "    \n",
    "    new_text = \"\"\"\n",
    "    \n",
    "    The output needs to be formated as the following: \n",
    "    \n",
    "    {\n",
    "    \"topic\": {\n",
    "    \"topic1\": [\"Keyword1\", \"Keyword2\", \"Keyword3\"],\n",
    "    \"topic2\": [\"Keyword1\", \"Keyword2\", \"Keyword3\"]\n",
    "    }\n",
    "    }\n",
    "    \n",
    "    Only output the dictionary above, nothing else with it.\n",
    "    \"\"\"\n",
    "\n",
    "    prompt = PromptTemplate(\n",
    "    template=\" So you are a text assistant and I need you to help me identify the topics from the following list the text given to you {topics}. \\n Here's the text: {abstract}. \\n\\n Note: A single text can belong to multiple topics, so please list all relevant topics. {format_instructions}\",\n",
    "    input_variables=[\"format_instructions\", \"abstract\", \"topics\"]\n",
    "    )\n",
    "\n",
    "    chain = prompt | Llm | parser\n",
    "    topics = chain.invoke({\"format_instructions\": new_text, \"abstract\": abstract, \"topics\": topics})\n",
    "    return list(topics.values())[0]\n",
    "\n",
    "\n",
    "print(get_topic(\"The development of high-performance batteries is crucial for the future of electric vehicles. The current generation of batteries are not able to provide the range and power required for long-distance travel. This project aims to develop new materials for batteries that can provide higher energy density and faster charging times.\"))\n",
    "\n",
    "def get_info(abstract:str = None, **kwargs):\n",
    "    Llm = Ollama(model='llama3', temperature=0.5)\n",
    "    if abstract is None:\n",
    "        raise ValueError(\"Abstract is required\")\n",
    "    \n",
    "    dico = {}\n",
    "    for key, question in kwargs.items():\n",
    "        print(key, question)\n",
    "        prompt = PromptTemplate(\n",
    "            template=\"So you are a text assistant and I want you to assist me by providing the following information: {question}. \\n\\n Here's the text: {abstract}. \\n\\n If the text doesn't contain any information about the topic given, output: 'N/A'\",\n",
    "            input_variables=[\"abstract\", \"question\"]\n",
    "        )\n",
    "        chain = prompt | Llm \n",
    "        info = chain.invoke({\"abstract\": abstract, \"question\": question})\n",
    "        dico[key] = info\n",
    "    print(dico)\n",
    "    return dico\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow orchestration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in dico:\n",
    "    dico[key].update({\"topic\":get_topic(dico[key][\"abstract\"])})\n",
    "    dico[key].update(get_info(dico[key][\"abstract\"],affiliation=\"What affiliations do the authors or characters in the text have?\",\n",
    "                        new_materials=\"Does the text mention any new materials or discoveries?\",\n",
    "                        screening_algorithms=\"Are there any screening algorithms or systematic procedures discussed in the text?\",\n",
    "                        ai_algorithms=\"Does the text reference any AI algorithms or methods related to artificial intelligence?\",\n",
    "                        workflow=\"Can you describe the workflow or process followed in the text?\",\n",
    "                        methods=\"Can you summarize the methods or approaches mentioned in the text?\",\n",
    "                        models=\"What models or frameworks are discussed or used in the text?\",\n",
    "                        funding=\"Does the text mention any funding sources or sponsors?\"))\n",
    "\n",
    "print(dico)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing in json file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "with open(\"output.json\", 'w') as file:\n",
    "    json.dump(dico, file)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
